### PR TITLE
Fixed failing test in HEASARC module, adding local test data

### PR DESCRIPTION
**Pull request by volodymyrss**
The test fails with astropy==5.0. Which indicates correctly that most of the module functionality breaks. 
There is some complexity in astropy with version check in 5.0, but I noticed everything works fine with the latest astropy `main` branch.
Meanwhile, and nevertheless, might be good to just always specify the unit, this will work with all relevant astropy versions.

I also update test according to some server-side change. 

I wonder how to prevent this sort of problem from emerging at some point.
 It would be good to have non-remote tests I guess. Only non-remotes ones are likely to continue passing?

I could just copy the entire  `astroquery/heasarc/tests/test_heasarc_remote_isdc.py` to  `astroquery/heasarc/tests/test_heasarc_remote_isdc.py` and add monkeypatches according to the [doc](https://astroquery.readthedocs.io/en/latest/testing.html#monkeypatching), is this the recommended way?

### CODE REVIEW

**Comment by keflavich** [comment]
Yes, I think copying over all the remote tests to a non-remote variant would be helpful - it would be great if you could do that!

The other thing we'd like to do is have remote tests that test that we get the right results but are robust against upstream changes.  So in this case, we might test that the number of results is `>=274` instead of `==274`, but even that isn't guaranteed to stay consistent (sometimes entries are consolidated).  Another option is to check that the data contain something useful - e.g., `assert row_information in table['column']` - something like that.

File: astroquery/heasarc/core.py
Comment (keflavich) [inline astroquery/heasarc/core.py:92]: is this a hack to survey the whole sky?  Is this what is done through the online interface, or just something that you've verified works?

Comment (volodymyrss) [reply astroquery/heasarc/core.py:92]: I was not the one to add this approach here. I did change the radius from 100000 to 361 in the previous PR.
The purpose of this is to get the list of possible columns. 
It was decided by the original author that this is the way to do it. It seems to work so far.

**Comment by volodymyrss** [comment]
> Yes, I think copying over all the remote tests to a non-remote variant would be helpful - it would be great if you could do that!

It just seems like almost complete code repetition. 
In fact, I would be using some other custom monkeypatch to collect the data to store anyway.
But ok, if there is no better way, I copy.
 
> The other thing we'd like to do is have remote tests that test that we get the right results but are robust against upstream changes. So in this case, we might test that the number of results is `>=274` instead of `==274`, but even that isn't guaranteed to stay consistent (sometimes entries are consolidated). 

Yeah, something like that, unfortunately. I could narrow-down the query a bit, but it's still not guaranteed, basically no way to be certain, unless the service promises this sort of consistency.

> Another option is to check that the data contain something useful - e.g., `assert row_information in table['column']` - something like that.

File: astroquery/heasarc/core.py
Comment (keflavich) [reply astroquery/heasarc/core.py:92]: oh, yes, you're right, my bad

**Comment by keflavich** [comment]
True, there would be some code repetition, but there is real value in storing the output data needed for the tests locally, and that's more than a simple copy-paste.

It would really be great if we could set up an automated system to do that (write only remote tests, but cache the results and run the tests locally), but I fear such a system, while elegant, would have too many failure corner cases to be worth the effort.

**Comment by volodymyrss** [comment]
> True, there would be some code repetition, but there is real value in storing the output data needed for the tests locally, and that's more than a simple copy-paste.
> It would really be great if we could set up an automated system to do that (write only remote tests, but cache the results and run the tests locally), but I fear such a system, while elegant, would have too many failure corner cases to be worth the effort.

As I understood it, it's still a complete repetition of the entire test code. I am be wrong.
I quite appreciate the value storing output data in the repository. But I was not sure if there is a way to achieve it efficiently without lot's of manual work.

So I made an attempt to keep the options for harvesting data, running tests remotely and locally by parallelization (i.e. keeping single test source, no repetition).
There is a description here and the attempt around: https://github.com/oda-hub/astroquery/blob/27ad069bc5dd9cb034a8bad531371d47c840bc7c/astroquery/heasarc/tests/parametrization.py#L9-L15

* As long as the test data is there, it should run exactly as the standard `astroquery` approach (depending on the `remote_data` marker).
* If test data is not there, and remote_data is allowed - the data will be generated. It must be copied and committed manually.

Hopefully the implementation is not too convoluted, actually it's hardly more than the advised `astroquery` monkeypatch. 

Please let me know what you think!
If that's too convoluted, I will extract it into an external tool.
 
I was just curious to make an attempt, I do not really suggest it's completely generic, and I hope it's sufficiently noninvasive.
It will work for all my cases.

Meanwhile, while applying this, the CI tests for this module started to run for the first time, and fail... I will get to that.

**Comment by keflavich** [comment]
This is brilliant - thanks!  Yes, I think if you can make this work (looks like you have, but I haven't reviewed carefully yet), this is the best solution!

Could you make the 'parametrization' code general and move it into the `utils/` subpackage so we can use it for all modules?  I think that would be best done as a separate PR.   Or we can just finish up this one and then move it to the general case later.

**Comment by volodymyrss** [comment]
> This is brilliant - thanks! Yes, I think if you can make this work (looks like you have, but I haven't reviewed carefully yet), this is the best solution!

Good to hear!
I hope there is no hidden issue though. It works here, I will think more.
 
> Could you make the 'parametrization' code general and move it into the `utils/` subpackage so we can use it for all modules? I think that would be best done as a separate PR. Or we can just finish up this one and then move it to the general case later.

I think I rather prefer the second option: that way I test everything carefully on one case, while also fixing the case, and them move it in another PR.
Right now, few tests still fail, but I suspect it might be because they were not really run before.

**Comment by volodymyrss** [comment]
Ok, I am done with tests, though I am not sure the approach is the best possible.

Concerning the parametrization, after this is done we can discuss more in another PR?

For example, maybe it's strange that the `[save]` test variation is going to be always skipped in CI. Maybe it's not a problem, since remote ones are also often skipped. Or there could be an option to deselect them. Or there could be a way to actually run them in another CI workflow, to rebuild all saved data when possible? Maybe that's too much.

A minor challenge I found here, is that sometimes, with different dependencies, for the same test the request was slightly different (some rounding/converting assumptions), even if the response is the same. 
* It could be addressed two variations of test data for these case. If `[save]` is only triggered when there are no test data at all, it's necessary to keep a bit more track of which data should be available. Then, there is a minor danger to store more data than is ever used. Maybe this can be inspected automatically - which files were used.
* In this PR I just modified the request in one case (the same strange `0,0` request) which lead to this condition. Here this request was kind of pathological anyway.
* Using an option like `--remote-data` could also have an effect on this process, but it's again more invasive.

File: README.rst
Comment (bsipocz) [inline README.rst:1]: Pls do not remove this file. If possible stash the add back commit into the one that removed it.

**Review by bsipocz — CHANGES_REQUESTED** [review]
I get an error for a missing file, and there are a few fixable issues (and some nitpick comments). Overall looks great, thanks!

```
E           FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astroquery-test-sqy_voa0/lib/python3.9/site-packages/astroquery/heasarc/tests/data/8fe6c881.dat'
```

**Comment by bsipocz** [comment]
@volodymyrss - to reflect to your question about local vs remote tests. Yes, currently we only fail builds on non-remote test failures. But we run the remote tests from cron (weekly) cron and also manually. E.g here is a list of modules with failing remotes https://github.com/astropy/astroquery/issues/2203, as you see we fixed the wast majority already, but unfortunately for `heasarc` it wasn't clear whom to ping.

File: astroquery/heasarc/tests/parametrization.py
Comment (volodymyrss) [reply astroquery/heasarc/tests/parametrization.py]: I personally find the black's version here quite readable, for such a nested structure. 
Of course no problem to adjust as you wish!
But I struggle to put it in 3-4 short and readable lines.  I made it to 10.
Could you please make a suggestion?

File: astroquery/heasarc/tests/parametrization.py
Comment (volodymyrss) [reply astroquery/heasarc/tests/parametrization.py:42]: Should I make a separate issue and delete this comment?

File: astroquery/heasarc/core.py
Comment (volodymyrss) [reply astroquery/heasarc/core.py:20]: Ok, sorry, I had a problem with some recent `minversion` elsewhere, so I misinterpreted this.

So you suggest I add `ASTROPY_LT_5_1` in `utils`? It will not be satisfied before 5.1 release, right? Is it ok that the `devastropy` test will fail?

File: README.rst
Comment (volodymyrss) [reply README.rst]: Hm, file restored. I already squashed. I will squash at the and too, right? Then an extra commit now does not matter.

**Comment by volodymyrss** [comment]
> @volodymyrss - to reflect to your question about local vs remote tests. Yes, currently we only fail builds on non-remote test failures. But we run the remote tests from cron (weekly) cron and also manually. 

Good to know! But only with one python, one platform, right? 
Of course remote tests test that the assumptions about the services did not break.
So you repeat this sort of PR regularly, as services and dependencies change?
Maybe it's not always obvious how to disentangle breaking services from breaking dependencies. It was not completely trivial in this `heasarc` PR.
 
For me, it would be very useful for me to know if at `heasarc` (and some other modules) is usable with recent astropy on a variety of platforms.

So I suppose it's still quite useful to add all possible responses for multi-platform multi-version mock tests?

So what do you think about adding this parametrization to utils? 
I think it might work (perhaps with small modifications) for most services. This will make the modules which use it more reliable, I hope.
For example, I suppose `heasarc` will no more break so easily as it did, right?

I could propose a change for some other ones for which I care. In part some of those from #2203.

> E.g here is a list of modules with failing remotes #2203, as you see we fixed the wast majority already, but unfortunately for `heasarc` it wasn't clear whom to ping.

Thanks for the information!

I am not from HEASARC, so I can not be fully responsible for this. 
But we supply data for one of the missions (INTEGRAL). Also, we maintain our own HEASARC site (already included in astroquery), which was a special kind of US contribution to INTEGRAL.
Distributing INTEGRAL data is our ([ISDC](https://www.isdc.unige.ch/)) core function, and in the recent years we started to advise people to use astroquery  to access it.
So perhaps I care about this module more than most users of HEASARC, and I would be happy to help when and if feasible.

File: astroquery/heasarc/core.py
Comment (volodymyrss) [reply astroquery/heasarc/core.py:20]: I changed as suggested and it does fail astropy dev, as expected.

**Comment by volodymyrss** [comment]
Should there be a changelog entry for heasarc? It's only tests - no functionality is added or changed.

As suggested by @bsipocz, I added `ASTROPY_LT_5_1` since it was not there. Should there a changelog entry for that?

File: astroquery/heasarc/tests/parametrization.py
Comment (keflavich) [reply astroquery/heasarc/tests/parametrization.py:42]: yes, and let's correct the `get_mockreturn` documentation at that URL at the same time (https://github.com/astropy/astroquery/blob/main/docs/testing.rst)

File: astroquery/heasarc/tests/parametrization.py
Comment (keflavich) [inline astroquery/heasarc/tests/parametrization.py:70]: This suggestive warning is a good idea, but we should also have it point to some narrative documentation explaining why

Comment (keflavich) [reply astroquery/heasarc/tests/parametrization.py:70]: also might want to remind us that if we're `cp`ing into the source tree, we want to `git add` the changes.

File: astroquery/heasarc/tests/parametrization.py
Comment (keflavich) [inline astroquery/heasarc/tests/parametrization.py:78]: Could we add a docstring here?  Something like:

```
If mode is not remote, patch `requests.Session` such that all `request`s will return either a saved response or a locally defined get-mockreturn.
```

File: astroquery/heasarc/tests/parametrization.py
Comment (keflavich) [inline astroquery/heasarc/tests/parametrization.py:61]: perhaps add a comment here noting that `_origina_request` is a monkeypatch-added attribute in `patch_get`

**Review by keflavich — APPROVED** [review]
I have some minor documentation requests, but this is a great improvement.

File: astroquery/heasarc/tests/parametrization.py
Comment (volodymyrss) [reply astroquery/heasarc/tests/parametrization.py:61]: ok, done.

File: astroquery/heasarc/tests/parametrization.py
Comment (volodymyrss) [reply astroquery/heasarc/tests/parametrization.py:70]: > This suggestive warning is a good idea, but we should also have it point to some narrative documentation explaining why

Ok, but documentation where?

Comment (keflavich) [reply astroquery/heasarc/tests/parametrization.py:70]: https://astroquery.readthedocs.io/en/latest/testing.html

Comment (volodymyrss) [reply astroquery/heasarc/tests/parametrization.py:70]: > also might want to remind us that if we're `cp`ing into the source tree, we want to `git add` the changes.

Ok, added.

Comment (volodymyrss) [reply astroquery/heasarc/tests/parametrization.py:70]: > https://astroquery.readthedocs.io/en/latest/testing.html

I was thinking about something like that, but then this is still in the heasarc module? Does it make sense to add documentation on module-specific (for now) code in such a generic place?
If I add it there, maybe then better to follow the other option you suggested, and put it to `utils` in this PR?
Or add the extra details and documentation in the other PR.
As you wish.

Comment (keflavich) [reply astroquery/heasarc/tests/parametrization.py:70]: ah, no, you're quite right - we can delay adding to the narrative docs until that future PR

File: astroquery/heasarc/tests/parametrization.py
Comment (volodymyrss) [reply astroquery/heasarc/tests/parametrization.py:79]: I added it, but I guess it can be expanded in the future PR.

File: astroquery/heasarc/tests/parametrization.py
Comment (volodymyrss) [reply astroquery/heasarc/tests/parametrization.py:70]: Ok, noted.

**Comment by volodymyrss** [comment]
Thanks! 

I made an additional optional mechanism for the developer to customly specify function generating key for the request, [here](https://github.com/oda-hub/astroquery/blob/789d93d3427233f40a40a0e2e1cbd00510fe50c2/astroquery/heasarc/tests/parametrization.py#L99-L103) used [here](https://github.com/oda-hub/astroquery/blob/789d93d3427233f40a40a0e2e1cbd00510fe50c2/astroquery/heasarc/tests/test_heasarc_remote_isdc.py#L77-L78).
It may be useful if the request may change with time, while still returning compatible result.

I guess this sort of situation should be considered with considerable caution. But there was one test like that in this module. It could have been removed too (or modified to send the same requests), but it's somehow useful as it is. 
This small additional change does not affect much the rest of the change.

@bsipocz please let me know what you think about the astropy==5.1 version and my other updates and comments on your requests!

File: astroquery/heasarc/core.py
Comment (bsipocz) [reply astroquery/heasarc/core.py:20]: No, I think we better hack it around temporarily until the 5.1 release is out.

As I see this is the only remaining issue, and that there are a few merge commits, so I go ahead, rebase the PR and a commit that address this variable and we're good to go. Ultimately we better fix the unit parsings as the issue came up recently in #2333, too

**Comment by bsipocz** [comment]
OK, so I've rebased this PR to remove the merge commits as well as to squash all the commits into one that add and remove and otherwise manipulate the added datasets to avoid unnecessary size up of the git repository.  However, I'm unable to push back to the branch of the PR as it was opened from under an org (maintainers get write access to branches of the PRs if a tickbox is checked at the time of opening a PR, however as far as I know, it only works for the cases when the PR is opened for forks under personal accounts rather than from under orgs.)

@volodymyrss  - there are two ways to work around, please let me know which one you prefer:
 - make another PR from a personal fork that I can edit. I'm happy to do this from the rebased brach, the commits will still have you as author.
 - force push the branch that I rebased back to this PR (https://github.com/bsipocz/astroquery/tree/heasarc-unit-issue_rebased), you or anyone with write access to the oda-hub org can do that. 
 - (you could consider to add some write access to the three of us maintainers to the org's fork, that case we could push back code to the branches of the PRs if it ever comes back. Some orgs choose to do this, others switched to open PRs from the personal forks instead).

**Comment by volodymyrss** [comment]
@bsipocz perfect, thank you!  I force-pushed the branch, and also gave you access to oda-hub/astroquery.

**Review by bsipocz — APPROVED** [review]

**Comment by bsipocz** [comment]
Thanks @volodymyrss!

**Comment by bsipocz** [comment]
@volodymyrss - could you have a look at these failing CI builds? e.g. https://github.com/astropy/astroquery/runs/5726304297?check_suite_focus=true

It all was passing before merging this, but I suspect there maybe a bug with the hashing part as now a few tests are failing with missing files, all the while they are fine for most of the version combinations, as well as I'm unable to reproduce the failure locally.
Thanks!

**Pull request merged**
